File: android/library/adapter_oss/src/main/java/com/antgroup/galacean/effects/GEPlayer.java
package com.antgroup.galacean.effects;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import java.util.Map;

public class GEPlayer extends FrameLayout {

    private final GEPlayerImpl mImpl;

    public GEPlayer(@NonNull Context context, GEPlayerParams params) {
        super(context);

        mImpl = new GEPlayerImpl(params);
    }

    public void loadScene(Callback callback) {
        if (mImpl != null) {
            mImpl.loadScene(getContext(), callback);
            View view = mImpl.getView();
            if (view != null) {
                addView(view, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            }
        }
    }

    public void play(int repeatCount, GEPlayer.Callback callback) {
        if (mImpl != null) {
            mImpl.play(repeatCount, callback);
        }
    }

    public void play(int fromFrame, int toFrame, int repeatCount, GEPlayer.Callback callback) {
        if (mImpl != null) {
            mImpl.play(fromFrame, toFrame, repeatCount, callback);
        }
    }

    public void pause() {
        if (mImpl != null) {
            mImpl.pause();
        }
    }

    public void resume() {
        if (mImpl != null) {
            mImpl.resume();
        }
    }

    public void stop() {
        if (mImpl != null) {
            mImpl.stop();
        }
    }

    public void destroy() {
        if (mImpl != null) {
            mImpl.destroy();
        }
    }

    public float getAspect() {
        if (mImpl != null) {
            return mImpl.getAspect();
        }
        return -1;
    }

    public int getFrameCount() {
        if (mImpl != null) {
            return mImpl.getFrameCount();
        }
        return -1;
    }

    public static class GEPlayerParams {
        public String url;
        public Bitmap downgradeImage;
        public Map<String, String> variables;
        public Map<String, Bitmap> variablesBitmap;
    }

    public interface Callback {
        void onResult(boolean success, String errorMsg);
    }
}

File: ios/GalaceanEffects/Sources/Player/GalaceanEffects/GEPlayer.m
//
//  GEPlayer.m
//  MarsNative
//
//  Created by changxing on 2023/11/16.
//  Copyright © 2023 Alipay. All rights reserved.
//

#import "GEPlayer.h"
#import "GEPlayerImpl.h"

@implementation GEPlayerParams
@end

@interface GEPlayer ()

@property (nonatomic, strong) GEPlayerImpl* impl;

@end

@implementation GEPlayer

- (instancetype)initWithParams:(GEPlayerParams *)params {
    if (self = [super init]) {
        self.impl = [[GEPlayerImpl alloc] initWithParams:params];
    }
    return self;
}

- (void)loadScene:(void(^)(bool success, NSString *errorMsg))callback {
    if (self.impl) {
        [self.impl loadScene:callback];
        UIView* view = [self.impl getView];
        if (view) {
            [self addSubview:view];
            [view setTranslatesAutoresizingMaskIntoConstraints:NO];
            [self addConstraint:[NSLayoutConstraint constraintWithItem:view
                                                             attribute:NSLayoutAttributeWidth
                                                             relatedBy:NSLayoutRelationEqual
                                                                toItem:self
                                                             attribute:NSLayoutAttributeWidth
                                                            multiplier:1.0
                                                              constant:0.0]];
            [self addConstraint:[NSLayoutConstraint constraintWithItem:view
                                                             attribute:NSLayoutAttributeHeight
                                                             relatedBy:NSLayoutRelationEqual
                                                                toItem:self
                                                             attribute:NSLayoutAttributeHeight
                                                            multiplier:1.0
                                                              constant:0.0]];
        }
    }
}

- (void)playWithRepeatCount:(int)repeatCount
                   Callback:(void(^ _Nullable)(bool success, NSString *errorMsg))callback {
    if (self.impl) {
        [self.impl playWithRepeatCount:repeatCount Callback:callback];
    }
}

- (void)playWithFromFrame:(int)fromFrame
                  ToFrame:(int)toFrame
              RepeatCount:(int)repeatCount
                 Callback:(void(^ _Nullable)(bool success, NSString *errorMsg))callback {
    if (self.impl) {
        [self.impl playWithFromFrame:fromFrame ToFrame:toFrame RepeatCount:repeatCount Callback:callback];
    }
}

- (void)pause {
    if (self.impl) {
        [self.impl pause];
    }
}

- (void)resume {
    if (self.impl) {
        [self.impl resume];
    }
}

- (void)stop {
    if (self.impl) {
        [self.impl stop];
    }
}

- (void)destroy {
    if (self.impl) {
        [self.impl destroy];
    }
}

- (float)getAspect {
    if (self.impl) {
        return [self.impl getAspect];
    }
    return -1;
}

- (int)getFrameCount {
    if (self.impl) {
        return [self.impl getFrameCount];
    }
    return -1;
}

@end
